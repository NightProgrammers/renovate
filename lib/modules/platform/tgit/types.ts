export interface TGitIssue {
  id: number;
  iid: number;
  title: string;
  labels?: string[];
}

export interface TGitComment {
  body: string;
  id: number;
}

export interface TGitUser {
  id: number;
  username: string;
}

export interface TGitMergeRequest {
  id: number;
  iid: number;
  title: string;
  state: string;
  source_branch: string;
  target_branch: string;
  description: string;
  merge_status: string;
  assignee?: TGitUser;
  reviewers?: TGitUser[];
  labels: string[];
  sha: string;
}

export interface TGitMergeRequestReview {
  id: number;
  iid: number;
  state:
    | 'empty'
    | 'approving'
    | 'approved'
    | 'change_required'
    | 'change_denied';
  reviewers: TGitUser[];
  labels: string[];
}

export interface UpdateMergeRequest {
  target_branch?: string;
  title?: string;
  assignee_id?: number;
  reviewer_ids?: number[];
}

export type MergeMethod = 'merge' | 'squash' | 'rebase';

export interface RepoResponse {
  archived: boolean;
  default_branch: string;
  template_repository: boolean;
  ssh_url_to_repo: string;
  http_url_to_repo: string;
  forked_from_project: string;
  merge_requests_enabled: boolean;
  path_with_namespace: string;
  merge_method?: MergeMethod;
}

export interface TGitUserStatus {
  state: 'active' | string;
}
