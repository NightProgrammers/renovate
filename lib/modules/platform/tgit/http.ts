import { logger } from '../../../logger';
import { TGitHttp } from '../../../util/http/tgit';
import type { TGitUserStatus } from './types';

export const tgitApi = new TGitHttp();

export async function getUserID(username: string): Promise<number> {
  const url = `users/${username}`;
  return (await tgitApi.getJson<{ id: number }>(url)).body.id;
}

export async function isUserActive(user: string): Promise<boolean> {
  try {
    const url = `users/${user}`;
    const userStatus = (await tgitApi.getJson<TGitUserStatus>(url)).body;
    return userStatus.state === 'active';
  } catch (err) {
    logger.warn({ err }, 'Failed to get user status');
    return false;
  }
}
