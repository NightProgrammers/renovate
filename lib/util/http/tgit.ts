import is from '@sindresorhus/is';
import { PlatformId } from '../../constants';
import { logger } from '../../logger';
import { ExternalHostError } from '../../types/errors/external-host-error';
import { resolveBaseUrl } from '../url';
import type { HttpHeaders, HttpResponse, InternalHttpOptions } from './types';
import { Http } from '.';

let baseUrl = 'https://git.code.tencent.com/api/v3';
export const setBaseUrl = (url: string): void => {
  baseUrl = url;
};

interface TGitInternalOptions extends InternalHttpOptions {
  body?: string;
}

export interface TGitHttpHeaders extends HttpHeaders {
  // 项目总数
  'x-total'?: string | undefined;
  //总页数
  'x-total-pages'?: string | undefined;
  //	每页的项目数
  'x-per-page'?: string | undefined;
  // 当前页面的索引(从1开始)
  'x-page'?: string | undefined;
  // 下一页的索引
  'x-next-page'?: string | undefined;
  // 上一页的索引
  'x-prev-page'?: string | undefined;
}

interface TGitPageNode {
  cur: number;
  prev?: number;
  next?: number;
}

export interface TGitHttpOptions extends InternalHttpOptions {
  paginate?: boolean;
  token?: string;
}

function parsePageNode(headers: TGitHttpHeaders): TGitPageNode {
  const cur = headers['x-page'];
  const next = headers['x-next-page'];
  const prev = headers['x-prev-page'];
  if (!cur) {
    return null;
  }

  const ret: TGitPageNode = {
    cur: parseInt(cur),
  };
  if (next) {
    ret.next = parseInt(next);
  }
  if (prev) {
    ret.prev = parseInt(prev);
  }

  return ret;
}

export class TGitHttp extends Http<TGitHttpOptions, TGitHttpOptions> {
  constructor(type: string = PlatformId.TGit, options?: TGitHttpOptions) {
    super(type, options);
  }

  protected override async request<T>(
    url: string | URL,
    options?: TGitInternalOptions & TGitHttpOptions
  ): Promise<HttpResponse<T>> {
    const opts = {
      baseUrl,
      ...options,
      throwHttpErrors: true,
    };

    try {
      const result = await super.request<T>(url, opts);
      logger.warn(result.body);
      logger.warn(result.headers);
      const pn = parsePageNode(result.headers);
      if (opts.paginate && is.array(result.body) && pn && pn.next) {
        const resolvedUrl = new URL(
          resolveBaseUrl(options?.baseUrl ?? baseUrl, url)
        );
        resolvedUrl.searchParams.set('page', pn.next.toString());
        resolvedUrl.href;

        const nextResult = await this.request(resolvedUrl.toString(), opts);
        if (is.array(nextResult.body)) {
          result.body.push(...nextResult.body);
        }
      }

      return result;
    } catch (err) {
      throw handleGotError(err, url, opts);
    }
  }
}

function handleGotError(
  err: any,
  url: string | URL,
  opts: TGitHttpOptions
): Error {
  if (err.statusCode === 404) {
    logger.trace({ err }, 'TGit 404');
    logger.debug({ url: err.url }, 'TGit API 404');
    throw err;
  }
  logger.debug({ err }, 'TGit API error');
  if (
    err.statusCode === 429 ||
    (err.statusCode >= 500 && err.statusCode < 600)
  ) {
    throw new ExternalHostError(err, PlatformId.Gitlab);
  }

  if (err.name === 'ParseError') {
    throw new ExternalHostError(err, PlatformId.Gitlab);
  }
  throw err;
}
