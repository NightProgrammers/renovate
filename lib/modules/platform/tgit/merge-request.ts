import { logger } from '../../../logger';
import { tgitApi } from './http';
import type {
  TGitMergeRequest,
  TGitMergeRequestReview,
  UpdateMergeRequest,
} from './types';

export async function getMR(
  repository: string,
  id: number
): Promise<TGitMergeRequest> {
  logger.debug(`getMR(${id})`);

  const url = `projects/${repository}/merge_request/${id}`;
  const mr = (
    await tgitApi.getJson<TGitMergeRequest & { source_commit: string }>(url)
  ).body;
  mr.sha = mr.source_branch;

  const cr = await getMergeRequstReview(repository, id);
  mr.reviewers = cr.reviewers;
  if (mr.merge_status === 'can_be_merged' && cr.state !== 'approved') {
    mr.merge_status = cr.state;
  }

  return mr;
}

export async function updateMR(
  repository: string,
  id: number,
  data: UpdateMergeRequest
): Promise<void> {
  logger.debug(`updateMR(${id})`);

  const url = `projects/${repository}/merge_request/${id}`;
  await tgitApi.putJson(url, {
    body: data,
  });
}

async function getMergeRequstReview(
  repository: string,
  id: number
): Promise<TGitMergeRequestReview> {
  logger.debug(`getReview(${id})`);

  const url = `projects/${repository}/merge_request/${id}/review`;
  const res = await tgitApi.getJson<TGitMergeRequestReview>(url);

  return res.body;
}
