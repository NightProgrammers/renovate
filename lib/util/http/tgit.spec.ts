import * as httpMock from '../../../test/http-mock';
import { PlatformId } from '../../constants';
import { EXTERNAL_HOST_ERROR } from '../../constants/error-messages';
import * as hostRules from '../host-rules';
import { TGitHttp, setBaseUrl } from './tgit';

hostRules.add({
  hostType: PlatformId.TGit,
  token: '123test',
});

const tgitApiHost = 'https://git.code.tencent.com';
const selfHostedUrl = 'https://git.woa.com';

describe('util/http/tgit', () => {
  let tgitApi: TGitHttp;

  beforeEach(() => {
    tgitApi = new TGitHttp();
    setBaseUrl(`${tgitApiHost}/api/v3/`);
    delete process.env.TGIT_IGNORE_REPO_URL;

    hostRules.add({
      hostType: PlatformId.TGit,
      token: 'abc123',
    });
  });

  afterEach(() => {
    jest.resetAllMocks();

    hostRules.clear();
  });

  it('paginates', async () => {
    httpMock
      .scope(tgitApiHost)
      .get('/api/v3/some-url')
      .reply(200, ['a'], {
        'X-Page': '1',
        'X-Next-Page': '2',
      })
      .get('/api/v3/some-url?page=2')
      .reply(200, ['b', 'c'], {
        'X-Page': '2',
        'X-Prev-Page': '1',
        'X-Next-Page': '3',
      })
      .get('/api/v3/some-url?page=3')
      .reply(200, ['d'], {
        'X-Page': '3',
        'X-Prev-Page': '2',
      });
    const res = await tgitApi.getJson('some-url', { paginate: true });
    expect(res.body).toHaveLength(4);

    const trace = httpMock.getTrace();
    expect(trace).toHaveLength(3);
    expect(trace).toMatchSnapshot();
  });

  it('attempts to paginate', async () => {
    httpMock.scope(tgitApiHost).get('/api/v3/some-url').reply(200, ['a'], {
      'x-page': '1',
      'x-total-pages': '1',
    });
    const res = await tgitApi.getJson('some-url', { paginate: true });
    expect(res.body).toHaveLength(1);

    const trace = httpMock.getTrace();
    expect(trace).toHaveLength(1);
    expect(trace).toMatchSnapshot();
  });
  it('posts', async () => {
    const body = ['a', 'b'];
    httpMock.scope(tgitApiHost).post('/api/v3/some-url').reply(200, body);
    const res = await tgitApi.postJson('some-url');
    expect(res.body).toEqual(body);
    expect(httpMock.getTrace()).toMatchSnapshot();
  });
  it('sets baseUrl', () => {
    expect(() => setBaseUrl(`${selfHostedUrl}/api/v3/`)).not.toThrow();
  });

  describe('fails with', () => {
    it('403', async () => {
      httpMock.scope(tgitApiHost).get('/api/v3/some-url').reply(403);
      await expect(
        tgitApi.get('some-url')
      ).rejects.toThrowErrorMatchingInlineSnapshot(
        `"Response code 403 (Forbidden)"`
      );
    });

    it('404', async () => {
      httpMock.scope(tgitApiHost).get('/api/v3/some-url').reply(404);
      await expect(
        tgitApi.get('some-url')
      ).rejects.toThrowErrorMatchingInlineSnapshot(
        `"Response code 404 (Not Found)"`
      );
    });

    it('500', async () => {
      httpMock.scope(tgitApiHost).get('/api/v3/some-url').reply(500);
      await expect(tgitApi.get('some-url')).rejects.toThrow(
        EXTERNAL_HOST_ERROR
      );
    });

    it('ParseError', async () => {
      httpMock.scope(tgitApiHost).get('/api/v3/some-url').reply(200, '{{');
      await expect(tgitApi.getJson('some-url')).rejects.toThrow(
        EXTERNAL_HOST_ERROR
      );
    });
  });
});
