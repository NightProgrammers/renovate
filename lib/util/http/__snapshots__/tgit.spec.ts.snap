// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util/http/tgit attempts to paginate 1`] = `
Array [
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate, br",
      "host": "git.code.tencent.com",
      "private-token": "abc123",
      "user-agent": "RenovateBot/0.0.0-semantic-release (https://github.com/renovatebot/renovate)",
    },
    "method": "GET",
    "url": "https://git.code.tencent.com/api/v3/some-url",
  },
]
`;

exports[`util/http/tgit paginates 1`] = `
Array [
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate, br",
      "host": "git.code.tencent.com",
      "private-token": "abc123",
      "user-agent": "RenovateBot/0.0.0-semantic-release (https://github.com/renovatebot/renovate)",
    },
    "method": "GET",
    "url": "https://git.code.tencent.com/api/v3/some-url",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate, br",
      "host": "git.code.tencent.com",
      "private-token": "abc123",
      "user-agent": "RenovateBot/0.0.0-semantic-release (https://github.com/renovatebot/renovate)",
    },
    "method": "GET",
    "url": "https://git.code.tencent.com/api/v3/some-url?page=2",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate, br",
      "host": "git.code.tencent.com",
      "private-token": "abc123",
      "user-agent": "RenovateBot/0.0.0-semantic-release (https://github.com/renovatebot/renovate)",
    },
    "method": "GET",
    "url": "https://git.code.tencent.com/api/v3/some-url?page=3",
  },
]
`;

exports[`util/http/tgit posts 1`] = `
Array [
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate, br",
      "host": "git.code.tencent.com",
      "private-token": "abc123",
      "user-agent": "RenovateBot/0.0.0-semantic-release (https://github.com/renovatebot/renovate)",
    },
    "method": "POST",
    "url": "https://git.code.tencent.com/api/v3/some-url",
  },
]
`;
